includes/class-wc-payments-webhook-processing-service.php

```diff
@@ -0,0 +1,511 @@
+	 * @var WC_Payments_Account
+	 */
+	private $account;
+
+	/**
+	 * WC Payments Remote Note Service.
+	 *
+	 * @var WC_Payments_Remote_Note_Service
+	 */
+	private $remote_note_service;
+
+	/**
+	 * WC_REST_Payments_Webhook_Controller constructor.
+	 *
+	 * @param WC_Payments_DB                  $wcpay_db            WC_Payments_DB instance.
+	 * @param WC_Payments_Account             $account             WC_Payments_Account instance.
+	 * @param WC_Payments_Remote_Note_Service $remote_note_service WC_Payments_Remote_Note_Service instance.
+	 */
+	public function __construct(
+		WC_Payments_DB $wcpay_db,
+		WC_Payments_Account $account,
+		WC_Payments_Remote_Note_Service $remote_note_service
+	) {
+		$this->wcpay_db            = $wcpay_db;
+		$this->account             = $account;
+		$this->remote_note_service = $remote_note_service;
+	}
+
+	/**
+	 * Process webhook event data.
+	 *
+	 * @param  array $event_body Body data of webhook request.
+	 *
+	 * @return void
+	 *
+	 * @throws Rest_Request_Exception
+	 */
+	public function process( array $event_body ) {
+
+		$event_type = $this->read_rest_property( $event_body, 'type' );
+
+		switch ( $event_type ) {
+			case 'charge.refund.updated':
+				$this->process_webhook_refund_updated( $event_body );
+				break;
+			case 'charge.dispute.created':
+				$this->process_webhook_dispute_created( $event_body );
+				break;
+			case 'charge.dispute.closed':
+				$this->process_webhook_dispute_closed( $event_body );
+				break;
+			case 'charge.dispute.funds_reinstated':
+			case 'charge.dispute.funds_withdrawn':
+			case 'charge.dispute.updated':
+				$this->process_webhook_dispute_updated( $event_body );
+				break;
+			case 'charge.expired':
+				$this->process_webhook_expired_authorization( $event_body );
+				break;
+			case 'account.updated':
+				$this->account->refresh_account_data();
+				break;
+			case 'wcpay.notification':
+				$note = $this->read_rest_property( $event_body, 'data' );
+				$this->remote_note_service->put_note( $note );
+				break;
+			case 'payment_intent.payment_failed':
+				$this->process_webhook_payment_intent_failed( $event_body );
+				break;
+			case 'payment_intent.succeeded':
+				$this->process_webhook_payment_intent_succeeded( $event_body );
+				break;
+			case 'invoice.upcoming':
+				WC_Payments_Subscriptions::get_event_handler()->handle_invoice_upcoming( $event_body );
+				break;
+			case 'invoice.paid':
+				WC_Payments_Subscriptions::get_event_handler()->handle_invoice_paid( $event_body );
+				break;
+			case 'invoice.payment_failed':
+				WC_Payments_Subscriptions::get_event_handler()->handle_invoice_payment_failed( $event_body );
+				break;
+		}
+	}
+
+	/**
+	 * Safely get a value from the REST request body array.
+	 *
+	 * @param array  $array Array to read from.
+	 * @param string $key   ID to fetch on.
+	 *
+	 * @return string|array|int
+	 * @throws Rest_Request_Exception Thrown if ID not set.
+	 */
+	public function read_rest_property( $array, $key ) {
```

Comment:
In the context of a webhook processing service we should probably avoid REST terminology and classes (exceptions in this case).

Renaming the method would be the easy part (ex. `read_webhook_property`), and I think we could throw a new type of exception here, which can be converted to a `Rest_Request_Exception` in the webhook controller. WDYT?

Reply:
Agreed. I will iterate on this.

Reply:
The latest version of the PR should have this already. It's almost ready for the review. I will double check again and request an official review next week.